toolbox: Update geometry on scrollbar style (transient) changed

irgolic opened the pull request:
On Mac, a mouse plugged in and 'Show scroll bars automatically based on mouse or trackpad' selected in General System Preferences, the `SH_ScrollBar_Transient` style hint will at first act as if the scrollbar is transient. After a few moments, it will correctly report that the scrollbar is not transient. This resulted in the toolbox being slightly obscured by the scrollbar.

This change stores the sizehint in the instance, and schedules a geometry update if it's changed, correctly updating the size of the toolbox.

ales-erjavec commented:
This does not seem right. The `updateGeometry` is a 'notifier' for `sizeHint` changes itself. So it depends on sizeHint being called by someone.

I think the proper solution is to hook into `QEvent.StyleChange` events delivered to the scroll bar itself to trigger the`updateGeometry`.

Can you try this:
```diff
diff --git a/orangecanvas/gui/toolbox.py b/orangecanvas/gui/toolbox.py
index c5f575bfd..c91fccfe3 100644
--- a/orangecanvas/gui/toolbox.py
+++ b/orangecanvas/gui/toolbox.py
@@ -14,13 +14,14 @@ from typing import NamedTuple, List, Iterable, Optional, Any, Callable
 from AnyQt.QtWidgets import (
     QWidget, QFrame, QSizePolicy, QStyle, QStyleOptionToolButton,
     QStyleOptionToolBox, QScrollArea, QVBoxLayout, QToolButton,
-    QAction, QActionGroup, QApplication, QAbstractButton, QWIDGETSIZE_MAX,
+    QAction, QActionGroup, QApplication, QAbstractButton, QScrollBar,
+    QWIDGETSIZE_MAX,
 )
 from AnyQt.QtGui import (
     QIcon, QFontMetrics, QPainter, QPalette, QBrush, QPen, QColor, QFont
 )
 from AnyQt.QtCore import (
-    Qt, QObject, QSize, QRect, QPoint, QSignalMapper
+    Qt, QObject, QSize, QRect, QPoint, QSignalMapper, QEvent
 )
 from AnyQt.QtCore import Signal, Property
 
@@ -285,6 +286,15 @@ class ToolBox(QFrame):
                           designable=True,
                           doc="Exclusive tabs")
 
+    class __ToolBoxScrollBar(QScrollBar):
+        #: Emitted when the scroll bar receives a StyleChange event
+        styleChange = Signal()
+
+        def changeEvent(self, event: QEvent) -> None:
+            if event.type() == QEvent.StyleChange:
+                self.styleChange.emit()
+            super().changeEvent(event)
+
     def __init__(self, parent=None, **kwargs):
         # type: (Optional[QWidget], Any)-> None
         super().__init__(parent, **kwargs)
@@ -303,6 +313,11 @@ class ToolBox(QFrame):
             horizontalScrollBarPolicy=Qt.ScrollBarAlwaysOff,
             widgetResizable=True,
         )
+        # Scroll bar's style hints are used in sizeHint so we need to monitor
+        # its changes.
+        sb = ToolBox.__ToolBoxScrollBar()
+        sb.styleChange.connect(self.updateGeometry)
+        self.__scrollArea.setVerticalScrollBar(sb)
         self.__scrollArea.setFrameStyle(QScrollArea.NoFrame)
 
         # A widget with all of the contents.

```

irgolic commented:
@ales-erjavec Thank you, this works great!

@markotoplak This appears to be the definitive solution to the problem. I remember you showing me some other solutions for this same problem elsewhere. For now, I've implemented it as Aleš suggested, but should this be moved to util and implemented elsewhere?
